cannot parse correctly escape sequences in the string

sample test: https://github.com/kyoh86/go-yaml/commit/54285c4d51ad6379854f0a2a49169250f5c2a765

failed result:
```
--- FAIL: TestDecoder (0.00s)
    --- FAIL: TestDecoder/"1":_"\"2\"" (0.00s)
        decode_test.go:935: failed to test ["1": "\"2\""], actual=[map[1:\]], expect=[map[1:"2"]]
    --- FAIL: TestDecoder/"1":_"\"" (0.00s)
        decode_test.go:935: failed to test ["1": "\""], actual=[map[1:\]], expect=[map[1:"]]
    --- FAIL: TestDecoder/"1":_"X\"z" (0.00s)
        decode_test.go:935: failed to test ["1": "X\"z"], actual=[map[1:X\]], expect=[map[1:X\"z]]
    --- FAIL: TestDecoder/"1":_"\\\\" (0.00s)
        decode_test.go:935: failed to test ["1": "\\\\"], actual=[map[1:\\\\]], expect=[map[1:\\]]
    --- FAIL: TestDecoder/"1":_"\\"2\\"" (0.00s)
        decode_test.go:935: failed to test ["1": "\\"2\\""], actual=[map[1:\\]], expect=[map[1:\"2\"]]
    --- FAIL: TestDecoder/"1":_"a\x2Fb\u002Fc\U0000002Fd" (0.00s)
        decode_test.go:935: failed to test ["1": "a\x2Fb\u002Fc\U0000002Fd"], actual=[map[1:a\x2Fb\u002Fc\U0000002Fd]], expect=[map[1:a/b/c/d]]
    --- FAIL: TestDecoder/"1":_"\a\b\t\n\v\f\r\e\#x20\"\/\\\N\_\L\P" (0.00s)
        decode_test.go:935: failed to test ["1": "\a\b\t\n\v\f\r\e\#x20\"\/\\\N\_\L\P"], actual=[map[1:\a\b\t\n\v\f\r\e\#x20\]], expect=[map[1:
            

  "/\��  ]]
FAIL
FAIL    github.com/goccy/go-yaml        0.010s

```

--- a/decode_test.go
+++ b/decode_test.go
@@ -315,11 +315,81 @@ func TestDecoder(t *testing.T) {
 			map[string]time.Time{"v": time.Date(2015, 2, 24, 18, 19, 39, 0, time.UTC)},
 		},
 
-		// Quoted values.
+		// Single Quoted values.
 		{
-			"'1': '\"2\"'",
-			map[interface{}]interface{}{"1": "\"2\""},
+			`'1': '2'`,
+			map[interface{}]interface{}{"1": `2`},
 		},
+		{
+			`'1': '"2"'`,
+			map[interface{}]interface{}{"1": `"2"`},
+		},
+		{
+			`'1': ''''`,
+			map[interface{}]interface{}{"1": `'`},
+		},
+		{
+			`'1': '''2'''`,
+			map[interface{}]interface{}{"1": `'2'`},
+		},
+		{
+			`'1': 'B''z'`,
+			map[interface{}]interface{}{"1": `B'z`},
+		},
+		{
+			`'1': '\'`,
+			map[interface{}]interface{}{"1": `\`},
+		},
+		{
+			`'1': '\\'`,
+			map[interface{}]interface{}{"1": `\\`},
+		},
+		{
+			`'1': '\"2\"'`,
+			map[interface{}]interface{}{"1": `\"2\"`},
+		},
+		{
+			`'1': '\\"2\\"'`,
+			map[interface{}]interface{}{"1": `\\"2\\"`},
+		},
+
+		// Double Quoted values.
+		{
+			`"1": "2"`,
+			map[interface{}]interface{}{"1": `2`},
+		},
+		{
+			`"1": "\"2\""`,
+			map[interface{}]interface{}{"1": `"2"`},
+		},
+		{
+			`"1": "\""`,
+			map[interface{}]interface{}{"1": `"`},
+		},
+		{
+			`"1": "X\"z"`,
+			map[interface{}]interface{}{"1": `X"z`},
+		},
+		{
+			`"1": "\\"`,
+			map[interface{}]interface{}{"1": `\`},
+		},
+		{
+			`"1": "\\\\"`,
+			map[interface{}]interface{}{"1": `\\`},
+		},
+		{
+			`"1": "\\\"2\\\""`,
+			map[interface{}]interface{}{"1": `\"2\"`},
+		},
+
+		/*
+			// TODO: Escape string
+			{
+				`"1": "a\x2Fb\u002Fc\U0000002Fd"`,
+				map[interface{}]interface{}{"1": `a/b/c/d`},
+			},
+		*/
 
 		{
 			"a: -b_c",

--- a/scanner/scanner.go
+++ b/scanner/scanner.go
@@ -211,6 +211,7 @@ func (s *Scanner) scanSingleQuote(ctx *Context) (tk *token.Token, pos int) {
 		if idx+1 < len(ctx.src) && ctx.src[idx+1] == '\'' {
 			// '' handle as ' character
 			value = append(value, c)
+			ctx.addOriginBuf(c)
 			idx++
 			continue
 		}

--- a/scanner/scanner.go
+++ b/scanner/scanner.go
@@ -193,30 +193,78 @@ func (s *Scanner) breakLiteral(ctx *Context) {
 	ctx.breakLiteral()
 }
 
-func (s *Scanner) scanQuote(ctx *Context, ch rune) (tk *token.Token, pos int) {
-	ctx.addOriginBuf(ch)
+func (s *Scanner) scanSingleQuote(ctx *Context) (tk *token.Token, pos int) {
+	ctx.addOriginBuf('\'')
 	startIndex := ctx.idx + 1
 	ctx.progress(1)
-	for idx, c := range ctx.src[startIndex:] {
+	src := ctx.src
+	size := len(src)
+	value := []rune{}
+	for idx := startIndex; idx < size; idx++ {
+		c := src[idx]
 		pos = idx + 1
 		ctx.addOriginBuf(c)
-		switch c {
-		case ch:
-			if ctx.previousChar() == '\\' {
-				continue
-			}
-			value := ctx.source(startIndex, startIndex+idx)
-			switch ch {
-			case '\'':
-				tk = token.SingleQuote(value, string(ctx.obuf), s.pos())
-			case '"':
-				tk = token.DoubleQuote(value, string(ctx.obuf), s.pos())
-			}
-			pos = len([]rune(value)) + 1
-			return
+		if c != '\'' {
+			value = append(value, c)
+			continue
 		}
+		if idx+1 < len(ctx.src) && ctx.src[idx+1] == '\'' {
+			// '' handle as ' character
+			value = append(value, c)
+			idx++
+			continue
+		}
+		tk = token.SingleQuote(string(value), string(ctx.obuf), s.pos())
+		pos = len([]rune(value)) + 1
+		return
 	}
 	return
+}
+
+func (s *Scanner) scanDoubleQuote(ctx *Context) (tk *token.Token, pos int) {
+	ctx.addOriginBuf('"')
+	startIndex := ctx.idx + 1
+	ctx.progress(1)
+	src := ctx.src
+	size := len(src)
+	value := []rune{}
+	for idx := startIndex; idx < size; idx++ {
+		c := src[idx]
+		pos = idx + 1
+		ctx.addOriginBuf(c)
+		if c == '\\' {
+			if idx+1 < size {
+				nextChar := src[idx+1]
+				switch nextChar {
+				case '"':
+					ctx.addOriginBuf(nextChar)
+					value = append(value, nextChar)
+					idx++
+					continue
+				case '\\':
+					ctx.addOriginBuf(nextChar)
+					idx++
+				}
+			}
+			value = append(value, c)
+			continue
+		} else if c != '"' {
+			value = append(value, c)
+			continue
+		}
+
+		tk = token.DoubleQuote(string(value), string(ctx.obuf), s.pos())
+		pos = len([]rune(value)) + 1
+		return
+	}
+	return
+}
+
+func (s *Scanner) scanQuote(ctx *Context, ch rune) (tk *token.Token, pos int) {
+	if ch == '\'' {
+		return s.scanSingleQuote(ctx)
+	}
+	return s.scanDoubleQuote(ctx)
 }
 
 func (s *Scanner) scanTag(ctx *Context) (tk *token.Token, pos int) {

